Geometry module, link and board mapping updates - 2

snwebb opened the pull request:
#### PR description:

Replacement/continuation of #459, after implementation of module rotation class in https://github.com/cms-sw/cmssw/pull/32684. Note that the changes from PR32684 have been copied here, as they do not currently exist in the PFCal-dev repository.
This adds new functionality to obtain the mapping between modules, lpgbts, and stage 1 and 2 FPGAs.
These are in the form of a `.json' file, which is then read-in and manipulated (json file only an example skeleton at present)

#### PR validation:

Tested that the code compiles and each function works in isolation.

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalModuleDetId.h:
```diff
@@ -0,0 +1,105 @@
+  bool isHScintillator() const { return (subdet() == HGCalHScTrigger); }
+  bool isForward() const { return true; }
+
+  //get u and v rotated to whole detector coordinates
+  int wholeDetectorU() const;
+  int wholeDetectorV() const;
```
Doesn't seem to be implemented. To be removed?

snwebb replied:
Thanks, removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+
+  unsigned sector0_mask_ = 0x7f;  // 7 bits to encode module number in 60deg sector
+
+  edm::FileInPath l1tModulesMapping_;
+  edm::FileInPath l1tLinksMapping_;
```
Not needed anymore

snwebb replied:
Removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+  std::move(tmp_vector.begin(), tmp_vector.end(), std::inserter(disconnected_modules_, disconnected_modules_.end()));
+  tmp_vector = conf.getParameter<std::vector<unsigned>>("DisconnectedLayers");
+  std::move(tmp_vector.begin(), tmp_vector.end(), std::inserter(disconnected_layers_, disconnected_layers_.end()));
+  //  geom_rotation_120_(HGCalGeomRotation::SectorType::Sector120Degree);
```
To be removed

snwebb replied:
Are you referring to the commented line (which I will remove) or the lines about the disconnected modules and layers. If the latter, why are they not relevant in this implementation?

jbsauvan replied:
Yes, I was just referring to the commented line

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+    HFNoseTriggerDetId trigger_cell_trig_id(trigger_cell_id);
+    tc_type = trigger_cell_trig_id.type();
+    layer = trigger_cell_trig_id.layer();
+    enum class WaferCentring { WaferCentred, CornerCentredY, CornerCentredMercedes };
```
Doesn't seem to be used. To be removed?

snwebb replied:
thanks, removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+}
+
+unsigned HGCalTriggerGeometryV9Imp3::getModuleSize(const unsigned module_id) const {
+  throw cms::Exception("FeatureNotImplemented") << "getModuleSize is not implemented in HGCalTriggerGeometryV9Imp3";
```
Should return 1. There is one wafer per module.

snwebb replied:
Done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+
+void HGCalTriggerGeometryV9Imp3::fillMaps() {
+  // read json mapping file
+  json mapping_config_;
```
Trailing underscore to be removed as it is not a class member

snwebb replied:
Removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+  json_input_file >> mapping_config_;
+
+  //Stage 1 to Stage 2 mapping
+  for (unsigned stage1_id = 0; stage1_id < mapping_config_["Stage1"].size(); stage1_id++) {
```
Here and below. Should we add some basic checks? For instance that the "Stage1" field is indeed existing in the json.

snwebb replied:
I have now added some `try,catch` statements for each block, which seems to work in testing.

snwebb replied:
This also enabled me to find a couple of bugs in this read-in, which are now fixed.

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+  return packed_value;
+}
+
+unsigned HGCalTriggerGeometryV9Imp3::packLayerModuleId(unsigned layer, unsigned module) const {
+  unsigned packed_value = 0;
+  unsigned subdet = ForwardSubdetector::HGCEE;
+  if (layer > heOffset_) {
+    layer -= heOffset_;
+    subdet = ForwardSubdetector::HGCHEF;
+  }
+  packed_value |= ((layer & HGCalDetId::kHGCalLayerMask) << HGCalDetId::kHGCalLayerOffset);
+  packed_value |= ((module & HGCalDetId::kHGCalWaferMask) << HGCalDetId::kHGCalWaferOffset);
+  packed_value |= ((subdet & DetId::kSubdetMask) << DetId::kSubdetOffset);
+  return packed_value;
+}
```
Should not be needed since the id scheme used here doesn't correspond to the new module ids.

snwebb replied:
Thanks, yes

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+    unsigned module = module_det_id_si.wafer();
+    unsigned layer = layerWithOffset(module_id);
+    module = (module & sector0_mask_);
+    links = links_per_module_.at(packLayerModuleId(layer, module));
```
The packing function used here is not the correct one

snwebb replied:
Changed - I am not sure if `layerWithOffset` still needs to be used here, and if so does the function itself need to be modified?

jbsauvan replied:
It depends on what is used to fill the `links_per_module_` map. The same packing function should be used here.

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+  json_input_file.close();
+}
+
+unsigned HGCalTriggerGeometryV9Imp3::packWaferId(int waferU, int waferV) const {
```
doesn't seem to be used

snwebb replied:
Removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+  return packed_value;
+}
+
+void HGCalTriggerGeometryV9Imp3::unpackWaferId(unsigned wafer, int& waferU, int& waferV) const {
```
Doesn't seem to be used

snwebb replied:
Removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+                                                                                                         : waferVAbs);
+}
+
+void HGCalTriggerGeometryV9Imp3::unpackLayerWaferId(unsigned wafer, unsigned& layer, int& waferU, int& waferV) const {
```
Doesn't seem to be used

snwebb replied:
thanks, removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+}
+
+HGCalGeomRotation::WaferCentring HGCalTriggerGeometryV9Imp3::getWaferCentring(unsigned layer) const {
+  if (layer <= 28) {  // CE-E
```
Can use `heOffset_`, which contains the number of EE layers

snwebb replied:
thanks, done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,938 @@
+
+bool HGCalTriggerGeometryV9Imp3::disconnectedModule(const unsigned module_id) const {
+  bool disconnected = false;
+  if (disconnected_modules_.find(HGCalDetId(module_id).wafer()) != disconnected_modules_.end())
+    disconnected = true;
```
Should be removed for the moment as it is not consistent with the new module id

snwebb replied:
Removed

snwebb commented:
-  Thanks for the comments - I've addressed the ones above (other than one which I had a query about).
-  I still need to update the json file
-  I get a crash when I run after the initialisation of this geometry file (see below) - would this be expected?

```
Begin processing the 1st record. Run 1, Event 409728, LumiSection 46 on stream 0 at 19-Feb-2021 15:23:48.927 CET
----- Begin Fatal Exception 19-Feb-2021 15:24:00 CET-----------------------
An exception of category 'Invalid DetId' occurred while
   [0] Processing  Event run: 1 lumi: 46 event: 409728 stream: 0
   [1] Running path 'ntuple_step'
   [2] Calling method for module HGCalTriggerNtupleManager/'hgcalTriggerNtuplizer'
Exception Message:
Cannot initialize HGCScintillatorDetId from b4a40091
```

snwebb commented:
Hi @jbsauvan - I have now added a realistic input json file (other than the Stage 1 to Stage 2 mapping which is preliminary).

jbsauvan commented:
Thanks @snwebb 
Concerning the `HGCScintillatorDetId` exception, it is due to the usage of `HGCScintillatorDetId` for scintillator modules in the ntuplizers.

I quickly updated the geometry tester class for `V9Imp3` and found some inconsistencies in the TC <-> module navigation (checking and comparing TCs from module and module from TC):
```
 Trigger cell 3003562773( EE:HSil= 1:0 type= 2 z= 1 layer= 1 wafer(u,v:x,y)= (-11,-5:17,-10) triggerCell(u,v:x,y)= (5,1:-14,9))
 has not been found in                                                                                                                            
 module HGCalModuleDetId::HFNose:EE:HSil:HScin=0:1:0:0 type= 2 z= -1 layer=1 sector=1 module(u,v)= (6,11)
 Available trigger cells are:                                                                         
      EE:HSil= 0:1 type= 1 z= -1 layer= 1 wafer(u,v:x,y)= (-6,0:12,0) triggerCell(u,v:x,y)= (4,5:9,-1)  
      EE:HSil= 0:1 type= 1 z= -1 layer= 1 wafer(u,v:x,y)= (-6,0:12,0) triggerCell(u,v:x,y)= (4,4:3,0)   
      EE:HSil= 0:1 type= 1 z= -1 layer= 1 wafer(u,v:x,y)= (-6,0:12,0) triggerCell(u,v:x,y)= (5,6:15,0)  
      EE:HSil= 0:1 type= 1 z= -1 layer= 1 wafer(u,v:x,y)= (-6,0:12,0) triggerCell(u,v:x,y)= (5,5:9,2)  
      EE:HSil= 0:1 type= 1 z= -1 layer= 1 wafer(u,v:x,y)= (-6,0:12,0) triggerCell(u,v:x,y)= (5,4:3,4)                  
      EE:HSil= 0:1 type= 1 z= -1 layer= 1 wafer(u,v:x,y)= (-6,0:12,0) triggerCell(u,v:x,y)= (6,7:21,2)    
      EE:HSil= 0:1 type= 1 z= -1 layer= 1 wafer(u,v:x,y)= (-6,0:12,0) triggerCell(u,v:x,y)= (6,6:15,4)  
      EE:HSil= 0:1 type= 1 z= -1 layer= 1 wafer(u,v:x,y)= (-6,0:12,0) triggerCell(u,v:x,y)= (6,5:9,6)                                                                                  
      EE:HSil= 0:1 type= 1 z= -1 layer= 1 wafer(u,v:x,y)= (-6,0:12,0) triggerCell(u,v:x,y)= (6,4:3,8)
.....
```
You can find the tester class in commit https://github.com/jbsauvan/cmssw/commit/c5a9dd2d4266e77e7555e2b2f35da497461d6015

Not sure where the issue is, but it seems there is some problem with the `zside` and wafer `u,v`

snwebb commented:
Thanks @jbsauvan . I found a few issues that contributed to the problem:

Firstly there was a one character error where there was a `U` rather than a `V`
Secondly the definition of the subdetector was not fully consistent throughout (I made it more like the `HGCalDetId`).
Thirdly the definition of the z-side (i.e. does 0 or 1 mean negative or positive) was opposite in `HGCalModuleDetId` and `HGCalTriggerDetId`. The reason for this is because the definitions in `HGCalDetId` (where I copied the function from) and `HGCalTriggerDetId` were already inverted.

snwebb commented:
I fixed a couple of issues:
-  I've set the `hSc_num_panels_per_sector_` to 8 (since there are 288 cells in 360º)
- There was a bug in the way I got the TC eta from the module - I've now put it back to what is was like for the V2Impl

The tests are passed in the tester program, although it crashes at the end with the following error:

```
----- Begin Fatal Exception 02-Mar-2021 19:28:05 CET-----------------------
An exception of category 'StdException' occurred while
   [0] Processing  stream begin Run run: 1 stream: 0
   [1] Calling method for module HGCalTriggerGeomTesterV9Imp3/'hgcaltriggergeomtester'
Exception Message:
A std::exception was thrown.
_Map_base::at
```

jbsauvan commented:
This is due to an issue in fetching the number of links in modules at 
https://github.com/PFCal-dev/cmssw/blob/47fb05ce1edb5ed1abf13a4bf1bbbc870dd2ec26/L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc#L537-L538

snwebb commented:
Thank you - I have now added "disconnected modules" (i.e. modules that currently have no lpGBT links) to the json config file. The tester class now runs without errors.

jbsauvan commented:
Thanks @snwebb 
I will update the tester class to be able to check also lpgbt and boards mappings.
Just one thing that can be changed in the meantime: the number of links per module should be the number of elinks and not lpgbts. The lpgbts are not directly connected to modules and this is the number of elinks that is important for instance for BestChoice.

snwebb commented:
Thanks - I have updated the code to take this into account.

jbsauvan commented:
Hi @snwebb 
I have checked in more details the lpgbt, stage-1 and stage-2 mappings.

## Stage-1 and lpgbt mappings
I have added some checks for the module <-> stage-1 mapping and found some issues [1]. The commit containing these updates is is https://github.com/jbsauvan/cmssw/commit/61d12bbee39d954a4071ca77ce8be95dbeb2aed5
I think we miss proper detids for lpgbts and FPGAs.
For instance, here:
https://github.com/PFCal-dev/cmssw/blob/1f37d1a7e3a727ce40213a48c1e6cafb7564ee1d/L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc#L591-L600

what is retrieved are the lpgbt numbers within one sector (as stored in the `module_to_lpgbts_` map), while the module id is defined for the whole HGCAL (`HGCalModuleDetId`).

So detids valid for the whole HGCAl should be defined for lpgbts as well as for Stage-1 FPGAs. Same thing for Stage-2 FPGAs, but that's a bit different in that case. Indeed Stage-2  FPGAs are covering two Stage-1 sectors, so they would probably need a different detid scheme compared to Stage-1 FPGAs.

## Stage-2 mapping
Currently in the json it seems that there is a number of Stage-2 FPGA equal to the TMT period (18). But we actually won't simulate the Time Multiplexing, since virtually all the TMT FPGAs will do the exact same thing but on different events. So I would remove the TMT duplicates in the json.

[1]
```
Checking module -> stage-1 -> module consistency                               
Error:                                                                                                                     
 Module 3003562773(HGCalModuleDetId::HFNose:EE:HSil:HScin=0:0:0:0 type= 0 z= -1 layer=3 sector=3 module(u,v)= (5,1))
 has not been found in                                                         
 stage-1 HGCalModuleDetId::HFNose:EE:HSil:HScin=0:1:0:0 type= 2 z= 1 layer=1 sector=1 module(u,v)= (6,11)
 Available modules are:                                                                                                       
 Connected lpgbts are:                                                    
                                                                                                     
%MSG-w HGCalTriggerGeometryTester:   HGCalTriggerGeomTesterV9Imp3:hgcaltriggergeomtester@streamBeginRun  11-Mar-2021 11:01:39 CET Run: 1 Stream: 0
Problem with the trigger geometry detected. Only the basic cells tree will be filled   
```

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,917 @@
+  geom_set getNeighborsFromTriggerCell(const unsigned) const final;
+
+  geom_set getStage1FpgasFromStage2Fpga(const unsigned) const final;
+  geom_set getLbgbtsFromStage1Fpga(const unsigned) const final;
```
Typo: should be `getLpgbtsFromStage1Fpga` (with a `p` instead of `b`)

snwebb commented:
Thanks @jbsauvan 
I've added a new class `HGCalBackendDetId` to handle lpGBTs and Stage 1 FPGAs as well as fixing various bugs. The tester runs fine now.
I still need to add something to handle the Stage 2 boards.

jbsauvan commented:
Hi @snwebb 
One more thing that I realize could be useful is a function to go directly from modules to Stage-1 FPGA, without the need to call manually module->lpgbt and lpgbt->stage-1 (as I did in the tester class). The best would probably be to fill an additional map module->stage-1 in the initialization, that can be directly accessed as the other maps.

snwebb commented:
Thanks @jbsauvan 

I have now added the structure to label the links between Stage1 and Stage2 FPGAs (which would anyway be needed I believe), and in this way obtain the mapping between the Stage1 and Stage2 FPGAs.

Stage2 FPGAs cover ~half the detector although there are 3 per end-cap (ignoring time multiplexing). Therefore I have kept the sector identifier also for Stage2, where the sector is taken as the one corresponding to the maximum overlap with Stage 1 (i.e. the lower value in phi). The sector for the Stage1 links is defined to be the same one as connected to the Stage1 FPGA (which might be different to the Stage2 FPGA it is connected to).

I have also added the direct module->stage-1 maps and associated function.

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalBackendDetId.h:
```diff
@@ -0,0 +1,70 @@
+  static const int kTypeLpGBT = 0;
+  static const int kTypeStage1FPGA = 1;
+  static const int kTypeStage1Link = 2;
+  static const int kTypeStage2FPGA = 3;
```
Would be more practical to define these inside an `enum`

snwebb replied:
thanks, done

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalBackendDetId.h:
```diff
@@ -0,0 +1,70 @@
+  static const int kHGCalSectorOffset = 11;
+  static const int kHGCalSectorMask = 0x3;
+  static const int kHGCalTypeOffset = 13;
+  static const int kHGCalTypeMask = 0x3;
```
The 2 bits used by the Type are currently completely used (4 values). I'm wondering if it would make sense to already add one more bit in case we want to encode more types in the future.

snwebb replied:
done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -670,48 +798,66 @@ void HGCalTriggerGeometryV9Imp3::fillMaps() {
+        lpgbt_to_modules_.emplace(lpgbt_id, packed_value);
+
+        //fill subsiduary module to stage 1 mapping
+        module_to_stage1_.emplace(packed_value, stage1_id);
```
There could be a silent bug if in the json file one module is split on two lpgbts and each lpgbt goes to different stage-1 fpgas by mistake (which shouldn't happen).
It should be checked that if `packed_value` is already in the map, the `stage1_id` should be the same.

snwebb replied:
thanks - I have added a check

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp3.cc:
```diff
@@ -649,6 +653,60 @@ bool HGCalTriggerGeomTesterV9Imp3::checkMappingConsistency() {
+
+    // Filling Stage 1 FPGA -> modules
+
+    edm::LogPrint("ModuleCheck") << "Checking module -> stage-1 -> module consistency";
```
Would it be possible to add also checks of the Stage1 <-> Stage2 mappings?

snwebb replied:
I have added this (but maybe could you confirm this is as you expected?)

jbsauvan replied:
Thanks, that looks good

jbsauvan commented on Geometry/HGCalCommonData/src/HGCalGeomRotation.cc:
```diff
@@ -135,8 +133,7 @@ unsigned HGCalGeomRotation::uvMappingTo60DegreeSector0(WaferCentring waferCentri
     }
-  }
-  if (moduleU >= moduleV && moduleV < 0) {
+  } else if (moduleU >= moduleV && moduleV < 0) {
```
Once the PR has been finalized, the updates/fixes in `HGCalCommonData` should be PRed into central CMSSW

snwebb replied:
I was already asked to do this (which is how I found out about the bug) - the PR is here: https://github.com/cms-sw/cmssw/pull/33191

snwebb commented:
Thanks a lot, I have implemented all of your suggestions.

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,1105 @@
+        if (module_to_stage1_.find(packed_value) == module_to_stage1_.end()) {
+          module_to_stage1_.emplace(packed_value, stage1_id);
+        } else {  //check that the stage1_id is the same as in the existing map
+          if (stage1_id != module_to_stage1_.at(packed_value)) {
```
Here a search in the map is done twice, either with `find()` +  `emplace()` or with `find()` + `at()`.
The same thing can be done with only one call to `emplace()` (avoiding two consecutive searches) as it inserts the element only if it doesn't already exists and it indicates whether the element has been inserted or not, as well as the iterator of the newly inserted or already existing element. See https://en.cppreference.com/w/cpp/container/unordered_map/emplace

snwebb replied:
thanks, done

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp3.cc:
```diff
@@ -0,0 +1,1397 @@
+    trigger_map_set stage2_to_stage1;
+    for (const auto& stage1 : stage1_to_modules) {
+      HGCalBackendDetId id(stage1.first);
+      HGCalTriggerGeometryBase::geom_set Stage2FPGAs = triggerGeometry_->getStage2FpgasFromStage1Fpga(id);
```
Names with leading capital are usually reserved for class names. Here `Stage2FPGAs` should rather be `stage2FPGAs`

snwebb replied:
changed

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp3.cc:
```diff
@@ -0,0 +1,1397 @@
+      HGCalBackendDetId stage2(stage2_modules.first);
+
+      // Check consistency of Stage-1 FPGA going to Stage 2 FPGA
+      HGCalTriggerGeometryBase::geom_set Stage1FPGAs = triggerGeometry_->getStage1FpgasFromStage2Fpga(stage2);
```
Same as above

snwebb replied:
changed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,1105 @@
+  bool disconnected = false;
+  if (disconnected_layers_.find(layerWithOffset(module_id)) != disconnected_layers_.end())
+    disconnected = true;
+  return disconnected;
```
It will be necessary to add a check for modules not linked to stage-1 FPGAs. Such modules should be considered as disconnected.

snwebb replied:
thanks - check added

jbsauvan commented:
After running some tests, I noticed that the current detid scheme chosen for `HGCalModuleDetId` cannot work, because it uses the same detector and subdetector types as `HGCalDetId`.
 It means that there is no way to distinguish a `HGCalModuleDetId` from a `HGCalDetId`. But  the other bits are organized differently; in particular the layer is not encoded in the same way in `HGCalDetId` and `HGCalModuleId`. So, for instance, when the layer number of a module is retrieved in `HGCalTriggerTools` [1], it cannot correctly get both `HGCalDetId` and `HGCalModuleId` layers.

What can be done is to still use the `Forward` detector type, but switch to `HGCTrigger` subdetector [2]. Which means that 4+3 bits are reserved for det and subdet [3], and only the remaining bits should be used for `HGCalModuleDetId` specific fields.
The same thing should be done for `HGCalBackendDetId`, and some bit should probably be used to distinguish `HGCalModuleDetId` and `HGCalBackendDetId`.


[1] 
https://github.com/PFCal-dev/cmssw/blob/9e67c404a1004576be6721af9acdf240cdbc0b91/L1Trigger/L1THGCal/src/HGCalTriggerTools.cc#L111
[2]
https://github.com/PFCal-dev/cmssw/blob/9e67c404a1004576be6721af9acdf240cdbc0b91/DataFormats/ForwardDetId/interface/ForwardSubdetector.h#L12
[3]
https://github.com/PFCal-dev/cmssw/blob/9e67c404a1004576be6721af9acdf240cdbc0b91/DataFormats/DetId/interface/DetId.h#L19-L22

snwebb commented:
Thanks @jbsauvan for the comments. I have edited the `HGCalModuleDetId` and `HGCalBackendDetId` class, such that the sub-detector is `HGCTrigger` and there is now an additional function `triggerSubdetId()` for `HGCalModuleDetId` to determine whether a module is silicon HF or HB, scintillator, or nose. There is also a bit which distinguishes the two classes as suggested.

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalBackendDetId.h:
```diff
@@ -58,8 +65,11 @@ class HGCalBackendDetId : public DetId {
+  static const int kHGCalClassIdentifierMask = 0x1;
 
   enum BackendType { LpGBT, Stage1FPGA, Stage1Link, Stage2FPGA };
+  enum HGCalClassIdentifier { ModuleDetId, BackendDetId };
```
This is common to `HGCalBackendDetId` and `HGCalModuleDetId` so the enum should be defined in an external header file. The enum should not be duplicated in the two detid classes.

snwebb replied:
I've added a new mini-class `HGCalBackendCommon` (not sure if this is the best way to do it)

jbsauvan replied:
A class is not needed. The enum can simply be defined as in `ForwardSubdetector.h`:
 https://github.com/PFCal-dev/cmssw/blob/hgc-tpg-devel-CMSSW_11_2_0_pre5/DataFormats/ForwardDetId/interface/ForwardSubdetector.h

snwebb replied:
thanks - I've modified it

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalModuleDetId.h:
```diff
@@ -33,12 +35,18 @@ class HGCalModuleDetId : public DetId {
   HGCalModuleDetId(uint32_t rawid);
   /** Constructor from subdetector, zplus, type, layer, sector, module numbers */
-  HGCalModuleDetId(ForwardSubdetector subdet, int zp, int type, int layer, int sector, int moduleU, int moduleV);
+  HGCalModuleDetId(int subdet, int zp, int type, int layer, int sector, int moduleU, int moduleV);
```
Would be better to use `HGCalTriggerSubdetector subdet` instead of `int subdet`

snwebb replied:
thanks, done

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalModuleDetId.h:
```diff
@@ -78,14 +89,18 @@ class HGCalModuleDetId : public DetId {
+  static const int kHGCalClassIdentifierOffset = 24;
+  static const int kHGCalClassIdentifierMask = 0x1;
+
+  enum HGCalClassIdentifier { ModuleDetId, BackendDetId };
```
Same comment as for `HGCalBackendDetId`

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -859,10 +844,10 @@ void HGCalTriggerGeometryV9Imp3::fillMaps() {
-          if (stage1_id != module_to_stage1_.at(packed_value)) {
+        auto result = module_to_stage1_.emplace(packed_value, stage1_id);
+        if (result.second == false) {  //check that the stage1_id is the same as in the existing map
+          //	  auto &ref_to_ddd = result.first->second;
```
Commented line to be removed

snwebb replied:
done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -859,10 +844,10 @@ void HGCalTriggerGeometryV9Imp3::fillMaps() {
+        auto result = module_to_stage1_.emplace(packed_value, stage1_id);
+        if (result.second == false) {  //check that the stage1_id is the same as in the existing map
+          //	  auto &ref_to_ddd = result.first->second;
+          if (stage1_id != result.first->second) {
```
This could be done in the first `if` statement:
```cpp
if(result.second == false && stage1_id != result.first->second)
```

snwebb replied:
done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -928,9 +913,9 @@ void HGCalTriggerGeometryV9Imp3::etaphiMappingFromSector0(int& ieta, int& iphi,
+  if (subdet == HGCalTriggerSubdetector::HGCalEETrigger) {  // CE-E
     return HGCalGeomRotation::WaferCentring::WaferCentred;
-  } else if (subdet == ForwardSubdetector::HGCHEF || subdet == ForwardSubdetector::HGCHEB) {
+  } else if (subdet == HGCalTriggerSubdetector::HGCalHSiTrigger) {
```
Is there a reason to have removed `HGCalHScTrigger`?

snwebb replied:
The reason is that the type of WafterCentring isn't really appropriate for`HGCalHScTrigger` (only refers to silicon layers) - but maybe there's a better way to catch this?

jbsauvan replied:
Maybe a warning message could be issued

snwebb replied:
thanks, I've added one

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -962,8 +947,15 @@ bool HGCalTriggerGeometryV9Imp3::validTriggerCell(const unsigned trigger_cell_id
   bool disconnected = false;
-  if (disconnected_layers_.find(layerWithOffset(module_id)) != disconnected_layers_.end())
+  HGCalModuleDetId id(module_id);
+  if (module_to_stage1_.find(packLayerSubdetWaferId(id.layer(), id.triggerSubdetId(), id.moduleU(), id.moduleV())) !=
```
Shouldn't it be the opposite? ( `module_to_stage1_.find(...) == module_to_stage1_.end()`)

snwebb replied:
Yes, thanks

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp3.cc:
```diff
@@ -1137,10 +1137,11 @@ void HGCalTriggerGeomTesterV9Imp3::fillTriggerGeometry()
     module_ = 0;
-    if (id.subdetId() == ForwardSubdetector::HGCHEB || id.subdetId() == ForwardSubdetector::HFNose) {
+    if (moduleSubdet_ == HGCalTriggerSubdetector::HGCalHScTrigger ||
+        moduleSubdet_ == HGCalTriggerSubdetector::HFNoseTrigger) {
```
I may have missed this before: `HFNose` is silicon, so it doesn't have `ieta`, `iphi`.

snwebb replied:
Thanks, removed.

jbsauvan commented:
Thanks @snwebb
Could you also update the `HGCalTriggerTools` to take into account `HGCalModuleDetId`?
In `layer()`, `isEm()`, `isSilicon()`, `zside()`.

snwebb commented:
Thanks a lot for the comments - I've performed the updates to `HGCalTriggerTools` as well.

jbsauvan commented:
Thanks @snwebb 
Now everything seems to run fine. I will make some final checks and go through the code one last time.

jbsauvan commented:
Hi @snwebb 
I've seen issues in the TC distributions using BestChoice, which is driven by the number of elinks. Looking back at the number of links distribution given by the geometry tester, the numbers don't seem to correspond to what is in the json file.
Could you check where the issue could come from?
Thanks!

snwebb commented:
Hi @jbsauvan 
Thanks for noticing - the number of elinks for the scintillator modules was hard-coded rather than taking the values from the json file. Could you check if everything is now as expected?

jbsauvan commented:
Actually, I see issues rather in the ECAL part. 

For instance here is the distribution of the number of elinks written in the output ntuple of the tester:
![image](https://user-images.githubusercontent.com/5569798/113867009-4b65d280-97ae-11eb-8422-d255319da857.png)

Here I looked at the modules with 0 elinks (red circles) function of the layer and radius:
![image](https://user-images.githubusercontent.com/5569798/113867934-6be25c80-97af-11eb-9094-4f1270aaf6ca.png)
There are much more modules with 0 links on the boundaries than expected.

Same thing looking in the first layer:
![image](https://user-images.githubusercontent.com/5569798/113868248-c4b1f500-97af-11eb-9ff0-2321cfd1a293.png)

snwebb commented:
Thanks @jbsauvan - I found a bug whereby the layer was being obtained from the incorrect module class. This is now fixed and the links distribution looks more sensible.

jbsauvan commented:
Thanks @snwebb !
That looks much better now. I checked TC distributions with Threshold, BestChoice and STC. 
They are all very close to `V9Imp2`, there are just small differences in the Scintillator part at low eta. e.g. here [1] for the Threshold option (which appeared with your latest scintillator change).
I suspect these are due to disconnected "modules" in the scintillator. Do we expect disconnected scintillator modules from the json?

[1] (red=V9Imp2, blue=V9Imp3)
![image](https://user-images.githubusercontent.com/5569798/114039813-5a6a8480-9883-11eb-9fe3-bd009e7536ac.png)
![image](https://user-images.githubusercontent.com/5569798/114039847-622a2900-9883-11eb-913f-45563804b0b1.png)

snwebb commented:
I've checked and these scintillator modules don't exist in the json file (so they are treated as disconnected - previously each module was hard-coded to have exactly 1 link). This may be expected, as I remember Paul saying that the scintillator section of his original file was filled 'by-hand' since only the silicon mapping was provided to him.

jbsauvan commented:
Do you know how the scintillator modules are defined in the json? Is it the same as in `getModuleFromTriggerCell()`? 
Could you check where the numbers of elinks in the scintillator come from? So far we have always been using 2 elinks per scintillator "module".

snwebb commented:
Yes the function `getModuleFromTriggerCell()` should work for the scintillator too (I basically copied the code there from the V2 with minimal changes).
The original list of scintillator modules I have been used is here. https://github.com/snwebb/hgcal-linkmapping/blob/master/data/FeMappingV7.txt#L4886-L5508
The 13th column is the number of trigger e-links, which may be 1, 2 or 3 it seems.

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalBackendCommon.h:
```diff
@@ -0,0 +1,10 @@
+#ifndef L1Trigger_L1THGCal_HGCalBackendCommon_H
+#define L1Trigger_L1THGCal_HGCalBackendCommon_H 1
+
+#include <iosfwd>
+#include "DataFormats/DetId/interface/DetId.h"
+#include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
```
These includes should not be needed in this file

snwebb replied:
The first include can be omitted, but the other two are necessary. I have however moved these to `HGCalModuleDetId.h` and `HGCalBackendDetId.h` which might make it cleaner/clearer.

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalBackendDetId.h:
```diff
@@ -0,0 +1,74 @@
+
+class HGCalBackendDetId : public DetId {
+public:
+  /** Create a null cellid*/
+  HGCalBackendDetId();
+  /** Create cellid from raw id (0=invalid tower id) */
+  HGCalBackendDetId(uint32_t rawid);
+  /** Constructor from zplus, type, sector, label */
+  HGCalBackendDetId(int zp, int type, int sector, int label);
+  /** Constructor from a generic cell id */
+  HGCalBackendDetId(const DetId& id);
+  /** Assignment from a generic cell id */
```
The comments don't seem to correspond to this class. There is no `cell` here

snwebb replied:
thanks, changed to `backend id`

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalModuleDetId.h:
```diff
@@ -0,0 +1,104 @@
+  HGCalModuleDetId(HGCalTriggerSubdetector subdet, int zp, int type, int layer, int sector, int moduleU, int moduleV);
+  /** Constructor from a generic cell id */
+  HGCalModuleDetId(const DetId& id);
+  /** Assignment from a generic cell id */
```
The comments don't seem to correspond to this class

snwebb replied:
thanks, changed to `module id`

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,1095 @@
+#include "DataFormats/ForwardDetId/interface/HGCSiliconDetIdToROC.h"
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "DataFormats/ForwardDetId/interface/HGCalTriggerDetId.h"
+#include "DataFormats/HcalDetId/interface/HcalDetId.h"
```
`HcalDetId` doesn't seem to be needed

snwebb replied:
removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,1095 @@
+#include "L1Trigger/L1THGCal/interface/HGCalBackendDetId.h"
+#include "DataFormats/ForwardDetId/interface/HFNoseDetIdToModule.h"
+
+#include "tbb/concurrent_unordered_set.h"
```
`tbb/concurrent_unordered_set` is not needed

snwebb replied:
removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,1095 @@
+
+#include "tbb/concurrent_unordered_set.h"
+#include <fstream>
+#include <iostream>
+#include <regex>
```
`iostream` and `regex` are not needed

snwebb replied:
removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,1095 @@
+  unsigned hSc_module_size_ = 12;  // in TC units (144 TC / panel = 36 e-links)
+  unsigned hSc_wafers_per_module_ = 3;
+
+  unsigned sector0_mask_ = 0x7f;  // 7 bits to encode module number in 60deg sector
```
Doesn't seem to be used

snwebb replied:
removed

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,1095 @@
+  unsigned subdet = module_det_id.triggerSubdetId();
+
+  unsigned links = 0;
+  // Scintillator
```
Comment to be changed

snwebb replied:
done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,1095 @@
+    links = 1;
+  }
+  // TO ADD HFNOSE : getLinksInModule
+  // Silicon
```
Comment to be changed: this is for both silicon and scintillator in HGC

snwebb replied:
done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,1095 @@
+  bool same_sector = stage1link_to_stage2_.at(id.label());
+  unsigned sector = id.sector();
+
+  if (same_sector == false) {
```
Would be shorter to write simply `if (!same_sector)`

snwebb replied:
done

jbsauvan commented on L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc:
```diff
@@ -0,0 +1,1095 @@
+  if (module_to_stage1_.find(packLayerSubdetWaferId(id.layer(), id.triggerSubdetId(), id.moduleU(), id.moduleV())) ==
+      module_to_stage1_.end()) {
+    disconnected = true;
+    return disconnected;
```
Not needed. The `return` at the end is enough

snwebb replied:
thanks, removed

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalBackendDetId.cc:
```diff
@@ -0,0 +1,44 @@
+#include "FWCore/Utilities/interface/Exception.h"
+#include <iostream>
+
+//#define DebugLog
```
Commented line to be removed

snwebb replied:
removed

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalModuleDetId.cc:
```diff
@@ -0,0 +1,50 @@
+#include "FWCore/Utilities/interface/Exception.h"
+#include <iostream>
+
+//#define DebugLog
```
Commented line to be removed

snwebb replied:
removed

jbsauvan commented on L1Trigger/L1THGCal/test/HGCalTriggerGeomTesterV9Imp3.cc:
```diff
@@ -0,0 +1,1397 @@
+  return true;
+}
+
+bool HGCalTriggerGeomTesterV9Imp3::checkNeighborConsistency() {
```
This function can actually be removed as there is no neighbor definition

snwebb replied:
thanks, removed

jbsauvan reviewed (commented):
@snwebb please find here remaining comments on tiny details

snwebb commented:
Hi @jbsauvan - thanks for the comments, which I have all implemented.

I have also added a new version of `hgcal_trigger_link_mapping_v1.json` using an updated definition of the scintillator module that Paul provided me. If this causes too many issues, then I can revert that change for now (but hopefully the changes are minor and this mapping is slightly more realistic).

jbsauvan commented:
Thanks @snwebb 
The TC distributions using the new json look fine, just a bit more different in the scintillator section (see e.g. [1] for BC with PU-driven allocation).

Now you can:
- run `scram b code-format`
- rebase on top of the latest devel branch `hgc-tpg-devel-CMSSW_11_3_0_pre6`, without the changes already in CMSSW (related to `HGCalGeomRotation`)
- and squash

[1] (red=V9Imp2, blue=V9Imp3), BC PU-driven in both cases
![image](https://user-images.githubusercontent.com/5569798/115719964-e351e780-a37c-11eb-921e-c69fc0d11332.png)

snwebb commented:
Thanks @jbsauvan , that's done.

jbsauvan commented on L1Trigger/L1THGCal/test/testHGCalL1TGeometryV11Imp3_cfg.py:
```diff
@@ -0,0 +1,122 @@
+# Other statements
+process.genstepfilter.triggerConditions=cms.vstring("generation_step")
+from Configuration.AlCa.GlobalTag import GlobalTag
+process.GlobalTag = GlobalTag(process.GlobalTag, 'auto:phase2_realistic', '')
```
Now the global tag to be used is `auto:phase2_realistic_T15`

snwebb replied:
thanks - updated

snwebb committed:
Geometry module, link and board mapping updates
